task/aontimer/src/main.rs

```diff
@@ -0,0 +1,75 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use aontimer as aon;
+use drv_riscv_plic_api::*;
+use userlib::*;
+
+task_slot!(INT_CONTROLLER, ext_int_ctrl);
+
+const AONTIMER_BARK: u32 = 0x1;
+const AONTIMER_TIME: u32 = 0x2;
+
+const BARK_INT: u32 = 0;
+
+fn sleep_and_listen(timer: &aon::AonTimer) {
+    sys_log!("Sleeping...");
+    let alarm = sys_get_timer().now;
+    sys_set_timer(Some(alarm + 1000), AONTIMER_TIME);
+    loop {
+        let result = sys_recv_closed(
+            &mut [],
+            AONTIMER_BARK | AONTIMER_TIME,
+            TaskId::KERNEL,
+        )
+        .unwrap();
+        if result.operation & AONTIMER_BARK != 0x0 {
+            sys_log!("{}", sys_get_timer().now);
+            //reset the interrupt
+            if timer.bark_cb.is_some() {
+                sys_log!("calling back");
+                (timer.bark_cb.unwrap())();
+            }
+        }
+        if result.operation & AONTIMER_TIME != 0x0 {
+            // Comment the following line of code to recieve the bark, and then be forced reset.
+            return; //You've woken up
+        }
+    }
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let int_ctrl = RiscvIntCtrl::from(INT_CONTROLLER.get_task_id());
+
+    int_ctrl.disable_int(BARK_INT).unwrap();
+
+    const AONTIMER_BASE: u32 = 0x4047_0000;
```

Comment:
AON timer base address is fixed for the opentitan chip. Why does this need to be supplied in the task as well? Shouldn't the driver have this available from chip.toml?

Reply:
This feature is in the queue, https://rivosinc.atlassian.net/browse/SW-431.  Looking upstream, it seems Oxide has always hard coded their peripheral addresses.

Reply:
Thanks for the pointer, Drew. Yeah, it seems weird that the address is being duplicated in chip.toml and also in the tasks. I see that upstream oxide uses some chip-level crate to pull in the addresses in the drivers which helps avoid the duplicate address in the tasks directly. 

My other point here was that I would expect the driver to be aware of the base address but not the task directly i.e. the hierarchy will be something like:
some task that wants to use timer -----> driver for the timer ----> interrupt controller

So, this way the tasks don't deal directly with internals of the timer. Instead, they rely on the driver and the driver knows about the base address and also deals with the interrupt controller.

Reply:
I linked this thread into the jira issue, and also added a TODO in the code.